File: algorithms/211.cpp
#include <iostream>
using namespace std;

class WordDictionary {
private:
    struct TrieNode {
        TrieNode *next[26];
        bool isWord;

        TrieNode() {
            isWord = false;
            for (int i = 0; i < 26; i++)
                next[i] = NULL;
        }
        ~TrieNode() {
            for (int i = 0; i < 26; i++)
                if (next[i])
                    delete next[i];
        }
    };

    TrieNode *trie = NULL;

    bool searchHelper(const string &word, int idx, TrieNode *p) {
        if (idx == word.size())
            return p != NULL && p->isWord;
        if (p == NULL)
            return false;
        if (word[idx] == '.') {
            for (int i = 0; i < 26; i++)
                if (searchHelper(word, idx+1, p->next[i]))
                    return true;
        }
        else {
            if (searchHelper(word, idx+1, p->next[word[idx]-'a']))
                return true;
        }
        return false;
    }
public:
    /** Initialize your data structure here. */
    WordDictionary() {
        trie = new TrieNode;
    }
    ~WordDictionary() {
        delete trie;
    }

    /** Adds a word into the data structure. */
    void addWord(string word) {
        TrieNode *p = trie, *pr = NULL;
        for (auto ch: word) {
            pr = p;
            p = p->next[ch-'a'];
            if (p == NULL) {
                p = new TrieNode;
                pr->next[ch-'a'] = p;
            }
        }
        p->isWord = true;
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return searchHelper(word, 0, trie);
    }
};

int main() {
    return 0;
}

File: algorithms/python3/211.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isWord = False

    def addWord(self, word):
        node = self
        for ch in word:
            if ch not in node.child:
                node.child[ch] = TrieNode()
            node = node.child[ch]
        node.isWord = True

    def search(self, word):
        child = self.child
        if len(word) == 0:
            return self.isWord
        elif word[0] == '.':
            return any(
                node.search(word[1:]) for node in child.values()
            )
        else:
            return word[0] in child and child[word[0]].search(word[1:])


class WordDictionary:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.trie = TrieNode()
        

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str2
        :rtype: void
        """
        self.trie.addWord(word)
        

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """
        return self.trie.search(word)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
